docs/_posts/2015-09-30-react-render-and-top-level-api.md

```diff
@@ -0,0 +1,133 @@
+---
+title: "React.render and the Top Level React API"
+author: jimandsebastian
+---
+
+
+When you're in React's world you are just building components that fit into other components. Everything is a component. Unfortunately not everything around you is built using React. At the root of your tree you still have to write some plumbing code to connect the outer world into React.
+
+The primary API for rendering into the DOM looks like this:
+
+```js
+ReactDOM.render(reactElement, domContainerNode)
+```
+
+To update the properties of an existing component, you call render again with a new element.
+
+If you're not doing a single page app you may need to plug in to an outer component system to do cleanup work. React will not automatically clean up a tree. You need to manually call:
+
+```js
+ReactDOM.unmountComponentAtNode(domContainerNode)
+```
+
+This is important and often forgotten. Forgetting to call `unmountComponentAtNode` will cause your app to leak memory. There is no way for us to automatically detect when it is appropriate to do this work. Every system is different.
+
+It is not unique to the DOM. If you want to insert a React Native view in the middle of an existing iOS app you will hit similar issues.
+
+## Helpers
+
+If you have multiple React roots, or a single root that gets deleted over time, we recommend that you always create your own wrapper API. These will all look slightly different depending on what your outer system looks like. For example, at Facebook we have a system that automatically ties into our page transition router to automatically call `unmountComponentAtNode`.
+
+It is an anti-pattern to call `ReactDOM.render` directly.
+
+In your environment you may want to always configure internationalization, routers, user data etc. If you have many different React roots it can be a pain to set up configuration nodes all over the place. By creating your own wrapper you can unify that configuration into one place.
+
+## Object Oriented Updates
+
+If you call `ReactDOM.render` a second time to update properties, all your props are completely replaced.
+
+```js
+ReactDOM.render(<App locale='en-US' userID={1} />, container);
+// props.userID == 1
+// props.locale == 'en-US'
+ReactDOM.render(<App userID={2} />, container);
+// props.userID == 2
+// props.locale == undefined ??!?
+```
+
+If you're integrating with an object oriented API around you this might be confusing. For example, if you're using React to build a Web Component (Custom Element). In object oriented programming, all state lives on each object instance and you apply changes incrementally by mutating that state, one at a time. Not all at once like React does.
```

Comment:
Two sentence fragments. I would probably just replace the whole paragraph with just

> In object-oriented programming, all state lives on each object instance and you apply changes incrementally by mutating that state, one piece at a time.